tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/beam_pipeline.py

```diff
@@ -0,0 +1,339 @@
+      element: Dict[Text, Dict[Text, tf.Tensor]],
+      spec: execution_spec.ExecutionSpec,
+      remote_op_name: Text
+  ) -> Iterator[Dict[Text, Dict[Text, tf.Tensor]]]:
+    """Executes a subgraph layer.
+
+    To execute a subgraph layer, we need to prepare a feed_dict by extracting
+    tensors from element. Then, we run the subgraph and store its outputs to
+    a copy of element.
+
+    Since we import `GraphDef` protos, all the node names now have the prefix
+    "import/". Also, TensorFlow feed_dict and outputs accept tensor
+    names instead of node names. Hence, a conversion from node_name to
+    "import/node_name:0" is necessary. Note that this conversion assumes
+    that there is one output per node.
+
+    Args:
+      element: A dictionary from remote op names to a dictionary from tensor
+               names to tensors. Element[remote_op_name] stores graph inputs
+               and previous specs' outputs.
+      spec: An ExecutionSpec for a subgraph layer.
+      remote_op_name: The remote op name of the current graph.
+
+    Returns:
+      A dictionary from remote op names to a dictionary from tensor names to
+      tensors. The dictionary is a copy of the input element, to which the
+      outputs of this subgraph layer have been added.
+    """
+    element = copy.deepcopy(element)
+    input_tensor_names = [_import_tensor_name(node_name)
+                          for node_name in spec.input_names]
+    output_tensor_names = [_import_tensor_name(node_name)
+                           for node_name in spec.output_names]
+    feed_dict = {tensor_name: element[remote_op_name][tensor_name]
+                 for tensor_name in input_tensor_names}
+
+    outputs = []
+    with tf.compat.v1.Session(graph=tf.Graph()) as sess:
+      tf.import_graph_def(spec.subgraph)
+      outputs = sess.run(output_tensor_names, feed_dict=feed_dict)
+
+    for output_tensor_name, output_tensor in zip(output_tensor_names, outputs):
+      element[remote_op_name][output_tensor_name] = output_tensor
+
+    yield element
+
+
+def _import_tensor_name(  # pylint: disable=invalid-name
+    node_name: Text) -> Text:
+  return 'import/%s:0' % node_name
+
+
+@beam.ptransform_fn
+@beam.typehints.with_input_types(Dict[Text, Dict[Text, tf.Tensor]])
+@beam.typehints.with_output_types(Dict[Text, Dict[Text, tf.Tensor]])
+def _LoadRemoteGraphInputs(  # pylint: disable=invalid-name
+    pcoll: beam.pvalue.PCollection,
+    parent_remote_op_name: Text,
+    child_remote_op_name: Text,
+    remote_op_name_to_graph_name: Mapping[Text, Text],
+    graph_to_remote_op_input_name_mapping:
+        Mapping[Text, Mapping[Text, Mapping[Text, Text]]]
+) -> beam.pvalue.PCollection:
+  """Loads remote graph inputs.
+
+  Before executing a remote graph, we need to prepare its inputs. We first
+  get the mapping from remote graph placeholder names to parent graph input
+  names. Then, in a copy of element, we load the remote graph inputs from
```

Comment:
Please define "load", or write a longer explanation (e.g. "copy the inputs from the parent graph's key to the remote graph's key in the copied dictionary")

Reply:
Changed.